tests/CallbacksTest.test.nut

```diff
@@ -0,0 +1,173 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO
+// EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+// OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+
+@include "./ConnectionManager.lib.nut"
+
+class CallbacksTest extends ImpTestCase {
+
+  cm = null;
+
+  function setUp() {
+    this.info("running setUp");
+    cm = ConnectionManager();
+  }
+
+  /*
+  * sets onDisconnect callback and disconnects device using CM
+  *
+  */
+  function testOnDisconnectAsync() {
+    return Promise(function(resolve, reject) {
+      cm.onDisconnect(function(expected) {
+        cm.connect();
+      }.bindenv(this));
+
+      cm.onConnect(function() {
+        resolve();
+      }.bindenv(this));
+
+      cm.disconnect();
+      assertTrue(!server.isconnected(), "should NOT be connected!");
+    }.bindenv(this))
+    .then(function(val) {
+        assertTrue(server.isconnected(), "should be connected again!");
+        _resetCM.call(this);
+    }.bindenv(this))
+    .fail(function(valueOrReason) {
+        _resetCM.call(this);
+        throw valueOrReason;
+      }.bindenv(this));
+  }
+
+ /*
+  * sets onConnect callback and disconnects device using CM
+  *
+  */
+  function testOnConnectAsync() {
+    local flag = 0;
+    return Promise(function(resolve, reject) {
+      cm.onDisconnect(function(expected) {
+        assertTrue(!flag, "should be 0 now");
+        cm.connect();
+      }.bindenv(this));
+
+      cm.onConnect(function() {
+        flag++;
+        resolve();
+      }.bindenv(this));
+
+      assertTrue(!flag, "should 0 before re-connect");
+      cm.disconnect();
+      assertTrue(!server.isconnected(), "should NOT be connected!");
+    }.bindenv(this))
+    .then(function(val) {
+      //not assertTrue(flag, ...) to ensure flag was increased exactly one time
+      assertTrue(flag==1, "connected again, should be 1 now");
+      _resetCM.call(this);
+    }.bindenv(this))
+    .fail(function(valueOrReason) {
+      _resetCM.call(this);
+      throw valueOrReason;
+    }.bindenv(this));
+  }
+
+ /*
+  * sets onNextConnect callback and disconnects device using CM
+  *
+  */
+  function testOnNextConnectAsync() {
+    local flag = 0;
+    return Promise(function(resolve, reject) {
+      cm.onDisconnect(function(expected) {
+        assertTrue(!flag, "should be 0 now");
```

Comment:
Do you have any reasons to put onNextConnect inside onDisconnect ? 

while onConnect method is not inside onDisconnect in the previous example.

Reply:
If you move out onNextConnect from the onDisconnect, then you can make an extra assert-check that onNextConnect doesn't call on or before disconnect();

Reply:
Yes, it was made to strictly order time when callback will be registered on action as a listener to allign invocation order.

Reply:
So, does it mean that if I register onNextConnect() when server.connected == true, then I can get callback immediately ? 

What does I mean:
local isNextConnectHappen = false;
cm.onDisconnect(function(expected) {
   assertTrue(!isNextConnectHappen, "On next connect should not happen before reconnect");
   cm.connect();
});
cm.onNextConnect(function(...) {
   isNextConnectHappen = true;
});

If you write down code in such way, then you'll be able to test one more use-case